### PR TITLE
[checkpoint] Stash device RNG state without the CUDA-style device-module API

### LINKED ISSUE #127676: module 'torch.mps' has no attribute 'device'
### 🐛 Describe the bug

I'm getting the error `module 'torch.mps' has no attribute 'device'` trying to train a `sentence-transformer model (which uses the huggingface transformers trainer).

The error is raised on this line -> https://github.com/pytorch/pytorch/blob/7ef7c265d4361691dc4cf54152db083de3215fbf/torch/utils/checkpoint.py#L183

Because the mps model doesn't have a `device()` method

https://github.com/pytorch/pytorch/blob/7ef7c265d4361691dc4cf54152db083de3215fbf/torch/mps/__init__.py

### Versions

Collecting environment information...
PyTorch version: 2.3.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.14 (main, May 25 2024, 13:44:20) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.3.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by 2sumtech**
Fixes #127676.

torch.utils.checkpoint saves the device RNG state during the forward and replays it during recompute, so a checkpointed function that draws randomness (dropout, rand_like) recomputes the same values and produces the same gradient. On MPS neither half happened, and the failure was silent. Backward recomputed the forward with a fresh dropout mask, so the gradient was simply wrong.

Two pieces of the legacy CUDA-shaped device-module contract are responsible. The stash is gated on `getattr(device_module, "_initialized", False)`, and torch.mps has no `_initialized`, so the gate was always False. Had it passed, `get_device_states` would have raised `AttributeError` anyway, since it selects the device with `device_module.device(idx)`, a context manager only torch.cuda and torch.xpu provide. Users hit that second one directly through libraries that stash RNG state themselves.

This PR removes checkpoint's dependency on both, instead of adding them to torch.mps (which is what the first version of this PR did).

- `get_device_states`/`set_device_states` now index the generator with `get_rng_state(device)` / `set_rng_state(state, device)`, exactly as `torch.random.fork_rng` (which checkpoint calls a few lines later on the same module) already does. No device context manager is needed here at all.
- The gate moves into `_is_device_initialized`, which reads `_initialized` when a backend has one (cuda/xpu/mtia unchanged, and a lazily initialized device is still never touched behind the user's back) and otherwise asks torch.accelerator whether this device type is the current accelerator. MPS has no lazy init step. cpu is never an accelerator and stays False as before.

I checked that `_is_device_initialized` returns exactly what the old `getattr` returned for cpu, cuda, xpu, mtia and meta, and only flips to True for mps. Public signatures of `get_device_states`/`set_device_states` are unchanged.

I did not route the device selection through `torch.accelerator.device_index`. `_accelerator_exchangeDevice` eagerly calls `maybe_initialize_device` and does not treat a negative index as a no-op the way `torch.cuda.device` does, so CUDA behavior would not have been identical for these public functions.

## Test Plan

The existing `test_activation_checkpoint_does_not_error` in test/test_mps.py now also compares the gradient of a dropout-style function checkpointed vs not, both `use_reentrant` values, fixed seed. On main that assertion fails with 28/64 elements mismatched. With this change it passes.

```
python test/test_mps.py TestMPS -k test_activation_checkpoint_does_not_error
```

On CPU the result set is identical before and after this change.

```
python test/test_utils.py -k checkpoint
python test/test_autograd.py -k checkpoint
```

Authored with AI assistance (Claude). I reviewed and tested the change on real Apple Silicon hardware.

### CODE REVIEW

File: torch/mps/__init__.py [hunk 1]
```diff
@@ -231,13 +256,26 @@ def _host_alias_storage(storage: "torch.UntypedStorage") -> "torch.UntypedStorag
...
+    # whenever the backend is available. Device-generic code such as
+    # torch.utils.checkpoint consults ``_initialized`` to decide whether to
+    # stash and restore the device RNG state. Expose it so checkpointed
+    # functions that use randomness (e.g. dropout) recompute with the same RNG
+    # state and produce correct gradients on MPS.
+    if name == "_initialized":
+        return is_available()
+    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
+
+
```
Comment (malfet) [inline torch/mps/__init__.py]: What's wrong with _initilized = True then? And why this is needed at all?

File: test/test_mps.py [hunk 2]
```diff
@@ -5140,6 +5140,52 @@ def fn(x):
...
+            x = torch.randn(64, device="mps", requires_grad=True)
+
+            torch.manual_seed(2024)
+            fn(x).backward()
+            ref_grad = x.grad.clone()
+
+            x.grad = None
+            torch.manual_seed(2024)
+            checkpoint(fn, x, use_reentrant=use_reentrant).backward()
+            self.assertEqual(x.grad, ref_grad)
```
Comment (malfet) [inline test/test_mps.py]: Why not integrate it into the test_activation_checkpoint_does_not_error above?

**Review by malfet — CHANGES_REQUESTED** [review]
A bit confused why __getattr__ is needed 
And lots of things changes since 2.3. Perhaps one should modify activation checkpointing to work with torch.accelerator API rather than implement some dummy classes? And if you still need that class, why not keep it to a minimum: Just class that have index=0 property wrapped in context manager decorator?

And be a bit specific with the test plan, you are adding 3 tests when activation checkpointing presumably already work (see tests I've mentioned)

**Comment by 2sumtech** [comment]
Fair points. Dropped the torch.mps additions entirely and fixed the caller instead, and folded the tests into the existing activation checkpoint test. One heads up, I went with reading the rng state per device instead of the accelerator device-index context manager, the why is in the PR description.
